### PR TITLE
Add aggregator_execution_enabled feature flag, and fallback logic

### PR DESCRIPTION
**Pull request by igor-aptos**
recreating as #9996 got closed as it was against Satya's PR.

### CODE REVIEW

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs [hunk 1]
```diff
@@ -203,30 +232,22 @@ fn native_snapshot(
...
-            )]))])
-        },
-        _ => Err(SafeNativeError::Abort {
-            abort_code: EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE,
-        }),
-    }
+    let (agg_value, _agg_max_value) = get_aggregator_fields_u128(&safely_pop_arg!(args, StructRef))?;
+
+    let result_value = if context.aggregator_execution_enabled() {
+        let aggregator_id: AggregatorID = aggregator_value_field_as_id(agg_value)?;
```
Comment (georgemitenkov) [inline aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:240]: We do not need type annotation here?

**Review by georgemitenkov — APPROVED** [review]
LGTM

File: aptos-move/framework/move-stdlib/sources/configs/features.move [hunk 2]
```diff
@@ -241,16 +241,26 @@ module std::features {
...
-        is_enabled(AGGREGATOR_SNAPSHOTS)
+    public fun aggregator_api_enabled(): bool acquires Features {
+        is_enabled(AGGREGATOR_API)
     }
 
+    const SAFER_RESOURCE_GROUPS: u64 = 31;
+
+    /// Whether the efficient and concurrent execution of aggregators is enabled.
+    /// Without this flags, aggregators are executed sequentially, creating
+    /// read/write conflicts.
```
Comment (vusirikala) [inline aptos-move/framework/move-stdlib/sources/configs/features.move:260]: Why would it create read/write conflicts if aggregators are disabled?

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs [hunk 3]
```diff
@@ -129,16 +127,30 @@ fn native_try_add(
...
-    let value = pop_value_by_type(&ty_args[0], &mut args)?;
-    let (id, limit) = pop_aggregator_fields_by_type(&ty_args[0], &mut args)?;
-    let aggregator = aggregator_data.get_aggregator(id, limit)?;
-    Ok(smallvec![Value::bool(
-        aggregator.try_add(aggregator_context.resolver, value)?
-    )])
+
+    let input = pop_value_by_type(&ty_args[0], &mut args)?;
+    let agg_struct = safely_pop_arg!(args, StructRef);
+    let (agg_value, agg_max_value) = get_aggregator_fields_u128(&agg_struct)?;
```
Comment (vusirikala) [inline aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:134]: Why is `pop_aggregator_fields_by_type` replaced by `get_aggregator_fields_u128` everywhere? 
We should pop either `u128` values of `u64` values depending on the input type. But `get_aggregator_fields_u128` will only pop out `u128` values independent of the input type, which is not accurate.

File: aptos-move/framework/move-stdlib/sources/configs/features.move [hunk 4]
```diff
@@ -241,16 +241,26 @@ module std::features {
...
+    /// After this flag is enabled, aggregators modifications stop creating conflict, as possible.
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/framework/move-stdlib/sources/configs/features.move:261]: I mean , there are still conflicts - reads create conflicts, incorrect branch prediction is equivalent to conflict , etc.

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs (hunk 3 above)
Comment (igor-aptos) [reply aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:134]: I mean even previous function was returning SafeNativeResult<(AggregatorID, u128)>, i.e. returning u128 for limit always - which is always OK - as it just widens the type. 
new function just returns u128 for value as well, because if execution is disabled - value can be u128, and cannot be safely cast to u64 to be put inside AggregatorID

Comment (vusirikala) [reply aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:134]: The older function pops either u64 or u128, and typecasts it to u128, and returns the u128. The `get_aggregator_fields_u128` function pops u128 even if the type is u64.

Comment (igor-aptos) [reply aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:134]: hm, I got confused with what which function does, you are right, updating

File: aptos-move/framework/src/natives/aggregator_natives/helpers_v1.rs [hunk 5]
```diff
@@ -39,6 +39,16 @@ pub(crate) fn get_aggregator_field(aggregator: &StructRef, index: usize) -> Part
     field_ref.read_ref()
 }
 
+/// Given a reference to `Aggregator` Move struct, updates a field value at `index`.
+pub(crate) fn set_aggregator_field(
```
Comment (vusirikala) [inline aptos-move/framework/src/natives/aggregator_natives/helpers_v1.rs:43]: This is helpers_v1 file for aggregators_v1. Why is this added now?

File: types/src/on_chain_config/aptos_features.rs [hunk 6]
```diff
@@ -37,9 +37,10 @@ pub enum FeatureFlag {
     EMIT_FEE_STATEMENT = 27,
     STORAGE_DELETION_REFUND = 28,
     SIGNATURE_CHECKER_V2_SCRIPT_FIX = 29,
-    AGGREGATOR_SNAPSHOTS = 30,
+    AGGREGATOR_API = 30,
     SAFER_RESOURCE_GROUPS = 31,
     SAFER_METADATA = 32,
+    AGGREGATOR_EXECUTION = 33,
```
Comment (vusirikala) [inline types/src/on_chain_config/aptos_features.rs:43]: AGGREGATOR_EXECUTION is 33 here. In features.move, it's 32. Is that fine?

Comment (igor-aptos) [reply types/src/on_chain_config/aptos_features.rs:43]: @vusirikala - nice catch, that was a rebase issue.

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs [hunk 7]
```diff
@@ -129,16 +151,30 @@ fn native_try_add(
...
     context.charge(AGGREGATOR_V2_TRY_ADD_BASE)?;
-    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
-    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
-    let value = pop_value_by_type(&ty_args[0], &mut args)?;
-    let (id, limit) = pop_aggregator_fields_by_type(&ty_args[0], &mut args)?;
-    let aggregator = aggregator_data.get_aggregator(AggregatorVersionedID::V2(id), limit)?;
-    Ok(smallvec![Value::bool(
-        aggregator.try_add(aggregator_context.resolver, value)?
-    )])
+
```
Comment (vusirikala) [inline aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:155]: There is no if-else statement checking aggregator_api_enabled flag.

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs [hunk 8]
```diff
@@ -150,16 +186,31 @@ fn native_try_sub(
...
     context.charge(AGGREGATOR_V2_TRY_SUB_BASE)?;
-    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
-    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
-    let value = pop_value_by_type(&ty_args[0], &mut args)?;
-    let (id, limit) = pop_aggregator_fields_by_type(&ty_args[0], &mut args)?;
-    let aggregator = aggregator_data.get_aggregator(AggregatorVersionedID::V2(id), limit)?;
-    Ok(smallvec![Value::bool(
-        aggregator.try_sub(aggregator_context.resolver, value)?
-    )])
+
```
Comment (vusirikala) [inline aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:190]: There is no if-else statement checking aggregator_api_enabled flag.

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs [hunk 9]
```diff
@@ -172,25 +223,28 @@ fn native_read(
...
     debug_assert_eq!(args.len(), 1);
-
     context.charge(AGGREGATOR_V2_READ_BASE)?;
-    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
-    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
-    let (id, limit) = pop_aggregator_fields_by_type(&ty_args[0], &mut args)?;
-    let aggregator = aggregator_data.get_aggregator(AggregatorVersionedID::V2(id), limit)?;
-    let value = aggregator.read_most_recent_aggregator_value(aggregator_context.resolver)?;
-    if value > limit {
+
```
Comment (vusirikala) [inline aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:227]: There is no if-else statement checking aggregator_api_enabled flag.

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs [hunk 10]
```diff
@@ -203,30 +257,23 @@ fn native_snapshot(
     mut args: VecDeque<Value>,
 ) -> SafeNativeResult<SmallVec<[Value; 1]>> {
     debug_assert_eq!(args.len(), 1);
-
     context.charge(AGGREGATOR_V2_SNAPSHOT_BASE)?;
-    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
-    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
```
Comment (vusirikala) [inline aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:261]: There is no if-else statement checking aggregator_api_enabled flag.

**Review by vusirikala — APPROVED** [review]

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs (hunk 7 above)
Comment (igor-aptos) [reply aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:155]: there is no need - we are gating constructors. 

when flag is disabled, nobody can create aggregators/snapshots, and so nobody can call this functions

after we enable the flag, and people create aggregators/snapshots, we can disable new construction, but shouldn't break contracts that created those. for that we have aggregator_execution_enabled flag which disables all complexities and should always work.
